medmodels/treatment_effect_estimation/builder.py

```diff
@@ -0,0 +1,312 @@
+from __future__ import annotations
+
+from typing import Any, Dict, Literal, Optional
+
+import medmodels.treatment_effect_estimation.treatment_effect as tee
+from medmodels.matching.algorithms.propensity_score import Model
+from medmodels.matching.metrics import Metric
+from medmodels.medrecord.querying import NodeOperation
+from medmodels.medrecord.types import (
+    Group,
+    MedRecordAttribute,
+    MedRecordAttributeInputList,
+)
+from medmodels.treatment_effect_estimation.analysis_modules.matching.matching import (
+    MatchingMethod,
+)
+
+
+class TreatmentEffectBuilder:
+    treatment: Group
+    outcome: Group
+
+    patients_group: Optional[Group]
+    time_attribute: Optional[MedRecordAttribute]
+
+    washout_period_days: Optional[Dict[str, int]]
+    washout_period_reference: Optional[Literal["first", "last"]]
+
+    grace_period_days: Optional[int]
+    grace_period_reference: Optional[Literal["first", "last"]]
+
+    follow_up_period_days: Optional[int]
+    follow_up_period_reference: Optional[Literal["first", "last"]]
+
+    outcome_before_treatment_days: Optional[int]
+
+    filter_controls_operation: Optional[NodeOperation]
+
+    matching_method: Optional[MatchingMethod]
+    matching_essential_covariates: Optional[MedRecordAttributeInputList]
+    matching_one_hot_covariates: Optional[MedRecordAttributeInputList]
+    matching_model: Optional[Model]
+    matching_distance_metric: Optional[Metric]
+    matching_number_of_neighbors: Optional[int]
+    matching_hyperparam: Optional[Dict[str, Any]]
+
+    def __init__(self) -> None:
+        pass
+
+    def set_treatment(self, treatment: Group) -> TreatmentEffectBuilder:
```

Comment:
these functions would more tipically be called `with_treatment`. Same goes for the other `set_` functions